Enhance Customer Email Password Reset include Login Username in Reset Password Link

alexookah opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR adds the username to the reset password link in WooCommerce. This enhancement is necessary for environments that utilize the REST API or GraphQL and require the username instead of the user ID for the password reset process.

### How to test the changes in this Pull Request:
Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Navigate to the Password Reset Page:
    - Go to `{{frontend url}}/my-account/lost-password/`

2. Request a Password Reset:
    - Enter a valid username or email address in the provided field.
    - Submit the request.

3. Verify Email Receipt:
    - Check your email inbox for a password reset email.
    - Confirm that the email has been received and contains a password reset link.

4. Inspect the Reset Link:
    - Ensure that the link in the email includes the username or email as a query parameter.
    - Verify that the link is correctly formatted and does not contain any errors.

5. Test the Password Reset Link:
    - Click on the reset link to be redirected to the password reset page.

6. Reset the Password:
    - Enter a new password in the provided fields.
    - Submit the new password.

7. Verify Password Update:
    - Confirm that the new password is accepted and the reset is successful.
    - Check for any error messages during the process.

8. Test Reset Link Without Username:
    - Use the password reset link from the email and manually remove the username query parameter.
    - Verify that you can still reset the password successfully and check for any errors or issues.



<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [x] Enhancement - Improvement to existing functionality

#### Message Add username in email reset-password link

</details>

<details>

<summary>Changelog Entry Comment</summary>
This PR enhances the password reset functionality by including the username in the reset password link, to be able to use from

</details>

Thank you for reviewing this PR! Please let me know if you need any further information or adjustments.

alexookah committed:
Add login username in reset password links

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

alexookah committed:
Move login param as final query param

barryhughes commented:
Hi @alexookah, thanks for submitting the PR!

> This enhancement is necessary for environments that utilize the REST API or GraphQL and require the username instead of the user ID for the password reset process.

Can you provide a little more detail here—how are you performing the reset?

alexookah commented:
Hi @barryhughes 
I am using wp graphql to reset the password. Which requires to use username. So if user receives a reset email password link it should contain also the username in the query params.

Source: https://www.wpgraphql.com/docs/users#reset-user-password

Let me know if you need some further clarification.

Looking forward for this to get merged 😃

barryhughes commented:
Thanks for the explanation. Agree this would also align things more with WP's default password reset link, which has the basic form (also, worth noting that the username is already included in the email, just not in the link):

```
site.url/wp-login.php?action=rp&key=<KEY>&login=<USERNAME>
```

barryhughes reviewed (changes requested):
1. Can you bump the version numbers for the templates you modified? Those should now be set to `9.3.0`.
2. You deleted the testing instructions segment of the PR template. Can you restore and populate? All PRs need testing instructions [(see notes here)](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions). In this instance, for example, we should at least describe the process for verifying this change has not broken the existing password reset flow.

alexookah committed:
Merge branch 'trunk' into fix_add_login_in_reset_pass_emails

alexookah committed:
Bump template versions to 9.3.0

alexookah commented:
Hello @barryhughes 
Thanks for the feedback!
I added in the PR description the testing instructions as you mention and also added the 9.3.0 woocommerce version.

Hope it is fine now for the merge :)

barryhughes commented on plugins/woocommerce/templates/emails/plain/customer-reset-password.php:
```diff
@@ -28,7 +28,7 @@
 echo sprintf( esc_html__( 'Username: %s', 'woocommerce' ), esc_html( $user_login ) ) . "\n\n";
 echo esc_html__( 'If you didn\'t make this request, just ignore this email. If you\'d like to proceed:', 'woocommerce' ) . "\n\n";
-echo esc_url( add_query_arg( array( 'key' => $reset_key, 'id' => $user_id ), wc_get_endpoint_url( 'lost-password', '', wc_get_page_permalink( 'myaccount' ) ) ) ) . "\n\n"; // phpcs:ignore
+echo esc_url( add_query_arg( array( 'key' => $reset_key, 'id' => $user_id, 'login' => rawurlencode( $user_login ) ), wc_get_endpoint_url( 'lost-password', '', wc_get_page_permalink( 'myaccount' ) ) ) ) . "\n\n"; // phpcs:ignore
```
Similar to earlier note, we should be specific about the linting rules we are ignoring:

```suggestion
echo esc_url( add_query_arg( array( 'key' => $reset_key, 'id' => $user_id, 'login' => rawurlencode( $user_login ) ), wc_get_endpoint_url( 'lost-password', '', wc_get_page_permalink( 'myaccount' ) ) ) ) . "\n\n"; // phpcs:ignore WordPress.Arrays.ArrayDeclarationSpacing.AssociativeArrayFound
```

barryhughes commented on plugins/woocommerce/templates/emails/customer-reset-password.php:
```diff
@@ -31,7 +31,7 @@
 <p><?php esc_html_e( 'If you didn\'t make this request, just ignore this email. If you\'d like to proceed:', 'woocommerce' ); ?></p>
 <p>
-	<a class="link" href="<?php echo esc_url( add_query_arg( array( 'key' => $reset_key, 'id' => $user_id ), wc_get_endpoint_url( 'lost-password', '', wc_get_page_permalink( 'myaccount' ) ) ) ); ?>"><?php // phpcs:ignore ?>
+	<a class="link" href="<?php echo esc_url( add_query_arg( array( 'key' => $reset_key, 'id' => $user_id, 'login' => rawurlencode( $user_login ) ), wc_get_endpoint_url( 'lost-password', '', wc_get_page_permalink( 'myaccount' ) ) ) ); ?>"><?php // phpcs:ignore ?>
```
```suggestion
	<a class="link" href="<?php echo esc_url( add_query_arg( array( 'key' => $reset_key, 'id' => $user_id, 'login' => rawurlencode( $user_login ) ), wc_get_endpoint_url( 'lost-password', '', wc_get_page_permalink( 'myaccount' ) ) ) ); ?>"><?php // phpcs:ignore WordPress.Arrays.ArrayDeclarationSpacing.AssociativeArrayFound ?>
```

We may as well take the opportunity to be specific about the linting rule we are ignoring (otherwise, we risk suppressing warnings about potentially problematic future changes).

barryhughes committed:
`phpcs:ignore` should be specific

barryhughes committed:
`phpcs:ignore` should be specific

barryhughes reviewed (approved):
Looks good to me, thanks!

- Made a small tweak to improve the (pre-existing) `phpcs:ignore` rule.
- Confirmed the username is present in the reset link, and that the password reset process works as before.
- Let's wait on CI for final confirmation, then we can merge.
